File: lua/5.3/starter.lua
local window = js.global
local document = window.document
local fkfig = require 'fkfig'
local paper = document:getElementById 'paper'
local paperContext = paper:getContext '2d'
local engine = fkfig.create()

local offsetX, offsetY, scale = 0, 0, 2

function resize()
  local width = engine.width
  local height = engine.height
  local iWidth = window.innerWidth
  local iHeight = window.innerHeight

  paper.width = iWidth
  paper.height = iHeight

  scale = math.floor(math.max(math.min(iWidth // width, iHeight // height), 1))
  offsetX = math.floor((iWidth - width * scale) / 2)
  offsetY = math.floor((iHeight - height * scale) / 2)
end
window:addEventListener('resize', resize)
resize()

function update(owner, tick)
  local g = paperContext

  engine:update(tick)
  engine:draw()

  g.imageSmoothingEnabled = false
  g:clearRect(0, 0, g.canvas.width, g.canvas.height)
  g:save()
  g:translate(offsetX, offsetY)
  g:scale(scale, scale)
  g:drawImage(engine.canvas, 0, 0)
  g:restore()

  if engine.running then
    window:requestAnimationFrame(update)
  end
end
engine:start()
window:requestAnimationFrame(update)
function window.stopGame()
  engine:stop()
end

File: starter.py
from figengine import FigEngine

from figshario import FigsharioMainMenu


def setup():
    cvs = document.getElementById('paper')
    ctx = cvs.getContext('2d')

    state = {
        'offset_x': 0,
        'offset_y': 0,
        'scale': 2
    }

    game = FigEngine()
    game.start()

    def resize():
        iwidth, iheight = window.innerWidth, window.innerHeight
        cvs.width = iwidth
        cvs.height = iheight

        scale = int(max(min(iwidth / game.width, iheight / game.height), 1))
        state['offset_x'] = (iwidth - game.width * scale) >> 1
        state['offset_y'] = (iheight - game.height * scale) >> 1
        state['scale'] = scale

    resize()
    window.addEventListener('resize', resize)

    def anim_frame(tick):
        if game.running:
            game.update(tick)

        if game.running:
            game.draw()

            scale = state['scale']
            ctx.save()
            ctx.translate(state['offset_x'], state['offset_y'])
            ctx.scale(scale, scale)
            ctx.drawImage(game.canvas, 0, 0)
            ctx.restore()

        if game.running:
            window.requestAnimationFrame(anim_frame)

    window.requestAnimationFrame(anim_frame)

    def stop_game():
        game.stop()

    window.stop_game = stop_game

    return game


def main():
    game = setup()
    scene = FigsharioMainMenu(game)
    game.push_scene(scene)


if __name__ == '__main__':
    main()
